shells/browser/shared/src/main.js

```diff
@@ -166,35 +166,35 @@ function createPanelIfReactLoaded() {
         // When the user chooses a different node in the browser Elements tab,
         // copy it over to the hook object so that we can sync the selection.
         chrome.devtools.inspectedWindow.eval(
-          '(window.__REACT_DEVTOOLS_GLOBAL_HOOK__.$0 = $0, undefined)',
-          (_, error) => {
+          '(window.__REACT_DEVTOOLS_GLOBAL_HOOK__.$0 !== $0) ?' +
+            '(window.__REACT_DEVTOOLS_GLOBAL_HOOK__.$0 = $0, true) :' +
+            'false',
+          (didSelectionChange, error) => {
             if (error) {
               console.error(error);
-            } else {
-              bridge.send('syncSelectionFromNativeElementsPanel');
+            } else if (didSelectionChange) {
+              // Remember to sync the selection next time we show Components tab.
+              needsToSyncElementSelection = true;
             }
           }
         );
       }
 
-      // When the user selects another item in the native Elements tab,
-      // select the corresponding React component.
-      let isListeningToNativeSelectionChange = false;
-      function ensureListeningToNativeSelectionChange() {
-        if (isListeningToNativeSelectionChange) {
-          return;
-        }
-        isListeningToNativeSelectionChange = true;
+      setReactSelectionFromBrowser();
+      chrome.devtools.panels.elements.onSelectionChanged.addListener(() => {
         setReactSelectionFromBrowser();
```

Comment:
I keep executing this eagerly because setting a field is super cheap and lets us have less of a delay when we actually do change the tab. The bridge event gets delayed until we show the tab tho.